### DISCUSSION TITLE
Encoders

### DISCUSSION THREAD

**Discussion by krystophny** [redmod-team/profit#160] [Ideas]
* encoders are currently a bit messy: input/output encoders and special cases to handle decoding the prediction variance
## Split Encoders ?
* raw encoder objects which just know `encode` and `decode` and act on arrays
* higher level objects which are aware of the underlying structure
  * labelled columns / variables - e.g. xarray
  * `encode_input`, `decode_variance` etc
* alternatively let this raw transformation be `class methods` of the encoder objects → can be used without instantiating objects

**Comment by Rykath** [comment]
## Discussion on 2022-03-18
* decoding the variance only works well with linear transformations (Normalization, PCA)
   * Could add method to these transformations and use it in surrogates. If it doesn't exist, `predict` should yield exception. 

* for non linear transformations it makes more sense to decode the confidence bounds
  * For plotting routines, surrogate should provide confidence bands directly as output instead of variance
  * Optional: Add term based on derivative of transformation

**Comment by Rykath** [comment]
For dimensionality reduction (PCA, KL) there should be an option to explicitly set the number of output dimensions / principal values as an alternative to the tolerance.
